### PR TITLE
cdh-web QA deploy playbook

### PR DESCRIPTION
**Pull request by thatbudakguy**
I'm currently able to run the playbook successfully start to finish, but `http://cdh-test-web1.princeton.edu/` times out (I am on VPN). You need to have the `community.general` ansible collection installed to run it; see the README for the command to do that. You can run the playbook with: `ansible-playbook playbooks/cdh-web_qa.yml`.

Tons of changes, since this involved touching virtually every role in some way, although for many of them it was just adding privilege escalation (and maybe a default or two). I also haven't messed with the tests for the `postgresql` role yet. Hoping more eyes on it will help clean up this massive PR.

### CODE REVIEW

**Comment by rlskoeser** [comment]
@thatbudakguy cdh-test-web1 is the server name, not the url — the ones we'd planned to use are documented here: 
https://github.com/pulibrary/princeton_ansible/issues/1798

— That reminds me, we'll need to update the allowed hostnames config.

File: roles/postgresql/tasks/main.yml
Comment (rlskoeser) [inline roles/postgresql/tasks/main.yml:21]: what are these needed for?

File: roles/postgresql/tasks/main.yml
Comment (rlskoeser) [inline roles/postgresql/tasks/main.yml:2]: I would like to see a comment mentioning & linking to the PUL postgresql role this is based on

File: roles/postgresql/tasks/backup_db.yml
Comment (rlskoeser) [inline roles/postgresql/tasks/backup_db.yml:9]: why admin? shouldn't app user be sufficient?

File: playbooks/cdh-web_qa.yml
Comment (rlskoeser) [reply playbooks/cdh-web_qa.yml:19]: Oh! I see it's part of the postgresql role, that makes a lot of sense.

File: roles/postgresql/tasks/backup_db.yml
Comment (rlskoeser) [inline roles/postgresql/tasks/backup_db.yml:13]: I think the current db backup puts content in tmp, but that's not ideal. Any problem with putting it under the deploy user somewhere? (could be a later improvement)

File: roles/postgresql/tasks/backup_db.yml
Comment (rlskoeser) [inline roles/postgresql/tasks/backup_db.yml:3]: How hard to skip the backup if create database resulted in a brand new db?

File: roles/deploy_user/molecule/default/verify.yml
Comment (rlskoeser) [inline roles/deploy_user/molecule/default/verify.yml:2]: update comment?

File: roles/postgresql/molecule/default/verify.yml
Comment (rlskoeser) [inline roles/postgresql/molecule/default/verify.yml:2]: Update stock comment.

Anything to understand here about the setup that made this work? The verify test looks good

File: roles/postgresql/molecule/default/verify.yml
Comment (rlskoeser) [inline roles/postgresql/molecule/default/verify.yml:24]: debug output still needed?

**Review by rlskoeser — COMMENTED** [review]
Some comments and questions, but generally I think this looks good. Was able to pull the latest version of the branch and run the molecule tests successfully.

You tested the new deploy, but might be good to test an old qa deploy also to make sure they still work.

File: group_vars/cdhweb_qa/vars.yml
Comment (thatbudakguy) [reply group_vars/cdhweb_qa/vars.yml:23]: This can be removed now, since the postgres branch was merged to develop - we're not using mysqlclient at all.

File: roles/postgresql/tasks/main.yml
Comment (thatbudakguy) [reply roles/postgresql/tasks/main.yml:21]: they're necessary to install the desired version of the postgresql client library in the next step.

File: roles/postgresql/tasks/backup_db.yml
Comment (thatbudakguy) [reply roles/postgresql/tasks/backup_db.yml:9]: i wasn't sure if executing the dump command was a special permission. we could try it and see!

File: roles/postgresql/tasks/backup_db.yml
Comment (thatbudakguy) [reply roles/postgresql/tasks/backup_db.yml:3]: hmm, great question. i think it would just be registering the output of the step that creates the db and checking if `changed: true`, maybe? that's how i would expect the ansible `postgresql` module to work, anyway.

File: roles/postgresql/molecule/default/verify.yml
Comment (thatbudakguy) [reply roles/postgresql/molecule/default/verify.yml:2]: It doesn't work - did it work for you? I wrote in the PR description that I was still messing with it - it isn't using Francis's recommended approach of installing postgres locally yet.

File: roles/postgresql/molecule/default/verify.yml
Comment (thatbudakguy) [reply roles/postgresql/molecule/default/verify.yml:24]: nope, good catch!

File: README.md
Comment (thatbudakguy) [reply README.md:73]: imo changelog makes sense in the context of releases but less sense without them, since you can just look at the list of commits. PUL has an interesting convention of [recording architecture-related decisions in markdown in the repo](https://github.com/pulibrary/princeton_ansible/tree/main/architecture-decisions), which seems sorta like our geniza decision log - maybe we could borrow that? not sure this would merit inclusion there though either.

File: roles/postgresql/tasks/backup_db.yml
Comment (rlskoeser) [reply roles/postgresql/tasks/backup_db.yml:9]: My expectation is that if you have read access to a database, you can dump it!

File: roles/postgresql/tasks/backup_db.yml
Comment (rlskoeser) [reply roles/postgresql/tasks/backup_db.yml:3]: that's my guess too

File: README.md
Comment (rlskoeser) [reply README.md:73]: Yes, I'm not sure either.  A decision log or date-based change log might be useful. The commit history isn't always intelligible enough for bigger picture stuff, especially over the long term.

File: roles/postgresql/molecule/default/verify.yml
Comment (rlskoeser) [reply roles/postgresql/molecule/default/verify.yml:2]: Oh, I misread the molecule output. No, it fails with an error when it's trying to pull the docker image.

Comment (thatbudakguy) [reply roles/postgresql/molecule/default/verify.yml:2]: Yeah, you'd need to build it locally first with `docker build -t pulpostgres .` in the directory containing the `Dockerfile`. That image doesn't exist on docker hub to pull down.

**Review by rlskoeser — APPROVED** [review]
Updates since last time all look good to me.

File: README.md
Comment (kayiwa) [reply README.md:73]: we do like the [architectural decisions](https://adr.github.io/) (I personally find it liberating) in that it allows me to be wrong and to return to the decision.

**Review by kayiwa — COMMENTED** [review]
Love your deploy user tests! 
postgresql looks good also.

File: README.md
Comment (thatbudakguy) [reply README.md:73]: Split out into #51.

**Comment by thatbudakguy** [comment]
@rlskoeser two things still need a decision before merging, I think:
- should we try to add tests for the postgres role? there still aren't any that use the PUL approach of running postgres locally to test against.
- should we try your suggestion of not making database backups if the database has only just been created? seems like low to medium effort, but we would also probably want to test it.

**Comment by rlskoeser** [comment]
@thatbudakguy thanks for identifying what's left and where you need me to weigh in.

At this point, I don't want to hold up the migration work any further.

> @rlskoeser two things still need a decision before merging, I think:
> 
> * should we try to add tests for the postgres role? there still aren't any that use the PUL approach of running postgres locally to test against.

I'd like to put this off. Would you make an issue to add a test? Can we set up a time down the road to pair with Francis on that, or borrow from the example hello world postrges role work he was doing for us?

> * should we try your suggestion of not making database backups if the database has only just been created? seems like low to medium effort, but we would also probably want to test it.

I don't think it's worth holding up the deploy, since it's pretty minor and not a huge downside (not a heavy cost for backing up an empty database). Should we make an issue to document it? I don't think it's a huge priority, but might be easy to fold in and test once we have the postgres role tested.

**Comment by thatbudakguy** [comment]
created #52 and #53 to track the above. for the postgres tests (#52), i think we know how to do it, so we can implement it and then tag francis for review.

File: roles/postgresql/molecule/default/verify.yml
Comment (thatbudakguy) [reply roles/postgresql/molecule/default/verify.yml]: Moved testing to #52.

File: roles/postgresql/tasks/backup_db.yml
Comment (thatbudakguy) [reply roles/postgresql/tasks/backup_db.yml:3]: tracked on #53.

**Pull request merged**
